lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs

```diff
@@ -15,28 +17,57 @@ module Cardano.Wallet.Primitive.AddressDiscovery.Random
     (
     -- ** State
       RndState (..)
-    , mkRndState
     ) where
 
 import Prelude
 
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..), Key (..), XPrv )
+import Cardano.Wallet.Primitive.AddressDerivation.Random
+    ( decryptDerPath, eskToHdPassphrase )
 import Cardano.Wallet.Primitive.AddressDiscovery
     ( CompareDiscovery (..)
     , GenChange (..)
     , IsOurs (..)
     , IsOwned (..)
     , KnownAddresses (..)
     )
-import Data.ByteString
-    ( ByteString )
+import Control.DeepSeq
+    ( NFData )
+import Crypto.Error
+    ( CryptoFailable (..) )
+import Data.Map.Strict
+    ( Map )
+import Data.Text
+    ( Text )
+import GHC.Generics
+    ( Generic )
 
-newtype RndState = RndState ByteString
+import qualified Data.Map.Strict as Map
 
-mkRndState :: RndState
-mkRndState = error "AddressDiscovery.Random unimplemented"
+-- | Unified identifier for wallets (both FO and EO ones).
+-- Technically it contains a text, but actually it contains one from two options:
+--
+-- 1. for FO-wallets - 'Address' made from wallet's root public key, in Base58-format.
+-- 2. for EO-wallets - 'UUID', in Base58-format.
+newtype HdRootId = HdRootId { getHdRootId :: Text }
+    deriving (Eq, Ord, Show, Generic)
+
+instance NFData HdRootId
+
+newtype RndState = RndState { getRndState :: Map HdRootId (Key 'RootK XPrv) }
+    deriving (Generic)
```

Comment:
Hmmm. Okay. So, there's a (few) major difference(s) between the old / legacy codebase and our new implementation. One of them is that the new implementation functions mostly work in the context of a *single wallet*. So, there's no such thing as a list or map of root private keys: a `WalletLayer` only knows about a single root private key at most.

So basicaly, I don't think we need anything more than the (single) root private key as a state for now. Hence: no need for a `HdRootId`.

Reply:
thanks for that remark, now `newtype RndState = RndState { getRndState :: (Key 'RootK XPrv) }`